src/mscorlib/src/System/Lazy.cs

```diff
@@ -202,27 +308,123 @@ public Lazy(Func<T> valueFactory, bool isThreadSafe)
+                if (ReferenceEquals(_state, executionAndPublication))
+                    return useDefaultConstructor ? ViaConstructor() : ViaFactory(LazyThreadSafetyMode.ExecutionAndPublication);
+
+                return Value;
+            }
+        }
+
+        private T PublicationOnly(LazyNonGeneric publicationOnly, T possibleValue)
+        {
+            try { }
+            finally
+            {
+                // we run this in a finally block to ensure that we don't get a partial completion due
+                // to a Thread.Abort, which could mean that other threads might be left in infinite loops
+                var previous = Interlocked.CompareExchange(ref _state, LazyNonGeneric.PublicationOnlySpinWait, publicationOnly);
+                if (previous == publicationOnly)
+                {
+                    _value = possibleValue;
+                    _factory = null;
+                    _state = null;
+                }
+            }
+
+            return Value;
+        }
+
+        private T PublicationOnlyViaConstructor(LazyNonGeneric initializer)
+        {
+            return PublicationOnly(initializer, CreateViaDefaultConstructor());
+        }
+
+        private T PublicationOnlyViaFactory(LazyNonGeneric initializer)
+        {
+            var factory = _factory;
+            if (factory == null)
+                return PublicationOnlySpinWait();
+            return PublicationOnly(initializer, factory());
+        }
+
+        private T PublicationOnlySpinWait()
+        {
+            var spinWait = new SpinWait();
+            while (!ReferenceEquals(_state, null))
+            {
+                // We get here when PublicationOnly temporarily sets _state to LazyNonGeneric.PublicationOnlySpinWait.
+                // This temporary state should be quickly followed by _state being set to null.
+                spinWait.SpinOnce();
+            }
+            return Value;
+        }
+
+        private T LazyGetValue(LazyNonGeneric state)
+        {
+            switch (state.State)
+            {
+                case LazyState.NoneViaConstructor:                    return ViaConstructor();
+                case LazyState.NoneViaFactory:                        return ViaFactory(LazyThreadSafetyMode.None);
+
+                case LazyState.PublicationOnlyViaConstructor:         return PublicationOnlyViaConstructor(state);
+                case LazyState.PublicationOnlyViaFactory:             return PublicationOnlyViaFactory(state);
+                case LazyState.PublicationOnlyWait:                   return PublicationOnlySpinWait();
+
+                case LazyState.ExecutionAndPublicationViaConstructor: return ExecutionAndPublication(state, true);
+                case LazyState.ExecutionAndPublicationViaFactory:     return ExecutionAndPublication(state, false);
+
+                default:
+                    state.ThrowException();
+                    return default(T);
+            }
```

Comment:
Majority of the initialization paths end with `return Value`. It may be better to make the initialization methods return void; and just have one `return Value` here - smaller code.